quint/src/runtime/impl/operatorEvaluator.ts

```diff
@@ -0,0 +1,167 @@
+// pop nargs computable values, pass them the 'fun' function, and
+
+import { Either, left, mergeInMany, right } from '@sweet-monads/either'
+import { Callable, Computable, EvaluationResult } from '../runtime'
+import { RuntimeValue, rv } from './runtimeValue'
+import { strict as assert } from 'assert'
+import { List } from 'immutable'
+import { QuintError, quintErrorToString } from '../../quintError'
+
+// push the combined computable value on the stack
+export function applyFun(
+  sourceId: bigint,
+  fun: (..._args: RuntimeValue[]) => EvaluationResult,
+  args: Computable[]
+): Computable {
+  // produce the new computable value
+  return {
+    eval: () => {
+      try {
+        // compute the values of the arguments at this point
+        const values = args.map(a => a.eval())
+        // if they are all defined, apply the function 'fun' to the arguments
+        return mergeInMany(values)
+          .mapLeft((errors): QuintError => ({ code: 'QNT501', message: errors.map(quintErrorToString).join('\n') }))
+          .chain(vs => fun(...vs.map(v => v as RuntimeValue)))
+      } catch (error) {
+        const msg = error instanceof Error ? error.message : 'unknown error'
+        return left({ code: 'QNT501', message: msg, reference: sourceId })
+      }
+    },
+  }
+}
+
+export function applyFold(order: 'fwd' | 'rev', args: Computable[]): Computable {
```

Comment:
I have re-written this implementation to use typescript's `reduce`/`reduceRight` instead of implementing fold by hand.